tests/test_xjse_calendar.py

```diff
@@ -0,0 +1,61 @@
+from unittest import TestCase
+import pandas as pd
+from pytz import UTC
+
+from .test_trading_calendar import ExchangeCalendarTestBase
+from trading_calendars.trading_calendar import WEEKENDS
+from trading_calendars.exchange_calendar_xjse import XJSEExchangeCalendar
+
+
+class XJSECalendarTestCase(ExchangeCalendarTestBase, TestCase):
+
+    answer_key_filename = 'xjse'
+    calendar_class = XJSEExchangeCalendar
+
+    # The JSE is open from 09:00 to 17:00.
+    MAX_SESSION_HOURS = 8
+
+    HAVE_EARLY_CLOSES = False
+
+    # Johannesburg does not use daylight savings time: South Africa
+    # Standard Time (SAST) is observed all year.
+
+    def test_no_weekend_sessions(self):
```

Comment:
I like the way you did these tests, but I would maybe add a couple more.  Two tests that I can think of off the top of my head are: 1) testing that holidays that fall on Sundays get made up on Monday, but holidays that fall on Saturday do not, and 2) testing ad hoc holidays.

Reply:
Adding a test for uniform Sunday-to-Monday logic seems good.

Since the ad-hoc holidays are just a static declaration, I think they're covered by testing the full set of 2019 holidays: if one of the ad-hoc holidays is present there, we can pretty safely assume they're working correctly.

Iterating through all the ad-hoc holidays to verify their presence would basically just be a re-statement of the `adhoc_holidays` definition, which usually isn't a very useful kind of test, and makes it annoying to make changes; but something like adding another full year's worth of holidays could help, since it's a different formulation of the definition ("the set of all holidays in a specific year", rather than "the set of statically defined holidays across all years").

(`test_no_weekend_sessions` might actually not be very useful itself, in that it's basically a re-statement of logic defined elsewhere. I think I'll merge it with the Sunday-to-Monday test.)